### PR TITLE
Updates to support national GHG method

### PR DESCRIPTION
**Pull request by bl-young**
New national method for GHG attribution, designed to mimic the [National GHG Industry Attribution Method](https://edg.epa.gov/metadata/catalog/search/resource/details.page?uuid=https://doi.org/10.23719/1517571). Includes major revisions to generation of GHG FBAs

Flowsa methodology updates:
- greater support for using a single crosswalk (or single source catalog name) across multiple FBAs including through new functions `return_true_source_catalog_name` and `check_activities_sector_like`
- improved handling of additional government sectors from BEA (e.g. S00101, S00202)
- update and rename to function `subset_BEA_table` to enable make or use table data as `ActivityConsumedBy` or `ActivityProducedBy` as data source. 
- new FBS parameters `clean_fba_before_mapping_df_fxn` and `source_flows` (to subset flow name in incoming FBA)

Other updates include:
- renaming of all NAICS crosswalk files
- crosswalk for USGS Mineral Yearbook for Lead (allocation source)
- implement use of yaml anchors in some existing FBS (no change in method)

### CODE REVIEW

**Comment by bl-young** [comment]
@catherinebirney @WesIngwersen opening this PR to discuss final validation of the national method. See current comparison in useeior to original dataset below.
GHG

**Comment by WesIngwersen** [comment]
Add @matthewlchambers to this review

**Comment by bl-young** [comment]
> Add @matthewlchambers to this review

Happy to have him take a look. I can't officially add him as a reviewer (as far as I can tell) since he does not have repo access

File: flowsa/flowbyfunctions.py [hunk 1]
```diff
@@ -265,10 +261,10 @@ def sector_aggregation(df_load, group_cols):
...
             df = df.append(agg_sectors, sort=False).reset_index(drop=True)
 
-    # manually modify non-NAICS codes that might exist in sector
-    # domestic/household
-    df = df.replace({'F0': 'F010',
-                     'F01': 'F010'})
+    # # manually modify non-NAICS codes that might exist in sector
+    # # domestic/household
+    # df = df.replace({'F0': 'F010',
+    #                  'F01': 'F010'})
```
Comment (catherinebirney) [inline flowsa/flowbyfunctions.py:267]: comment can be deleted @bl-young

**Review by catherinebirney — APPROVED** [review]
@bl-young In the fbs_yaml_refrac branch I'm modifying all code that takes the string length of a sector to instead read in the naics sector length crosswalk. Because manipulating any dfs with the household and government codes can't be based on string length. The code changes in that branch should be useful to ensure the ghg for government/household are allocated correctly

**Comment by catherinebirney** [comment]
> > Add @matthewlchambers to this review
> 
> Happy to have him take a look. I can't officially add him as a reviewer (as far as I can tell) since he does not have repo access

@WesIngwersen @matthewlchambers @bl-young Matthew has been added as a FLOWSA collaborator, so can now be tagged for reviews :)

**Comment by WesIngwersen** [comment]
> @catherinebirney @WesIngwersen opening this PR to discuss final validation of the national method. See current comparison in useeior to original dataset below. 

@bl-young I see we're very close, but do you have a good understanding of the differences that are present?

**Comment by WesIngwersen** [comment]
I would like to understand where were are with the [activity to sector mapping for the GHGI](https://github.com/USEPA/flowsa/blob/ghg_fbs/flowsa/data/activitytosectormapping/NAICS_Crosswalk_EPA_GHGI.csv)
Was the attempt here just to map to the same sectors as done in the original NGIAM, or have some mappings been corrected?

**Comment by WesIngwersen** [comment]
I get this fba related error when running getFlowBySector() when I still need a lot of the required FBAs stored locally
```
2022-01-11 09:10:00 INFO     BEA_Make_Detail_BeforeRedef 2012 not found in C:/Users/wesle/AppData/Local/flowsa/FlowByActivity/, running functions to generate FBA
2022-01-11 09:10:00 INFO     Creating dataframe list
Traceback (most recent call last):
  File "C:\Users\wesle\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-9f2d0322d50f>", line 1, in <module>
    GHG_2016_m1 = flowsa.getFlowBySector("GHG_national_2016_m1")
  File "c:\users\wesle\flowsa\flowsa\__init__.py", line 120, in getFlowBySector
    download_FBAs_if_missing=download_FBAs_if_missing
  File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 263, in main
    aset_names, download_FBA_if_missing)
  File "c:\users\wesle\flowsa\flowsa\fbs_allocation.py", line 123, in dataset_allocation_method
    **fba_dict)
  File "c:\users\wesle\flowsa\flowsa\fbs_allocation.py", line 476, in load_map_clean_fba
    **kwargs_dict
  File "c:\users\wesle\flowsa\flowsa\flowbyfunctions.py", line 850, in load_fba_w_standardized_units
    datasource, year, **fba_dict).reset_index(drop=True)
  File "c:\users\wesle\flowsa\flowsa\__init__.py", line 66, in getFlowByActivity
    flowsa.flowbyactivity.main(year=year, source=datasource)
  File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 219, in main
    source=source, year=year, config=config)
  File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 103, in call_urls
    confirm_gdrive = config.get('url').get('confirm_gdrive')
AttributeError: 'str' object has no attribute 'get'
```
Any help would be appreciated

**Comment by catherinebirney** [comment]
> I get this fba related error when running getFlowBySector() when my local machine still needs to create a lot of the required FBAs
> 
> ```
> 2022-01-11 09:10:00 INFO     BEA_Make_Detail_BeforeRedef 2012 not found in C:/Users/wesle/AppData/Local/flowsa/FlowByActivity/, running functions to generate FBA
> 2022-01-11 09:10:00 INFO     Creating dataframe list
> Traceback (most recent call last):
>   File "C:\Users\wesle\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
>     exec(code_obj, self.user_global_ns, self.user_ns)
>   File "<ipython-input-4-9f2d0322d50f>", line 1, in <module>
>     GHG_2016_m1 = flowsa.getFlowBySector("GHG_national_2016_m1")
>   File "c:\users\wesle\flowsa\flowsa\__init__.py", line 120, in getFlowBySector
>     download_FBAs_if_missing=download_FBAs_if_missing
>   File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 263, in main
>     aset_names, download_FBA_if_missing)
>   File "c:\users\wesle\flowsa\flowsa\fbs_allocation.py", line 123, in dataset_allocation_method
>     **fba_dict)
>   File "c:\users\wesle\flowsa\flowsa\fbs_allocation.py", line 476, in load_map_clean_fba
>     **kwargs_dict
>   File "c:\users\wesle\flowsa\flowsa\flowbyfunctions.py", line 850, in load_fba_w_standardized_units
>     datasource, year, **fba_dict).reset_index(drop=True)
>   File "c:\users\wesle\flowsa\flowsa\__init__.py", line 66, in getFlowByActivity
>     flowsa.flowbyactivity.main(year=year, source=datasource)
>   File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 219, in main
>     source=source, year=year, config=config)
>   File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 103, in call_urls
>     confirm_gdrive = config.get('url').get('confirm_gdrive')
> AttributeError: 'str' object has no attribute 'get'
> ```
> 
> Any help would be appreciated

@WesIngwersen latest commit addresses this error

**Comment by WesIngwersen** [comment]
@catherinebirney I made it a lot further this time into the fbs.. then ran into another config related error
```
2022-01-11 09:38:20 INFO     Retrieving Flow-By-Activity for datasource EPA_GHGI_T_3_42 in year 2016
2022-01-11 09:38:20 INFO     EPA_GHGI_T_3_42 2016 not found in C:/Users/wesle/AppData/Local/flowsa/FlowByActivity/, running functions to generate FBA
2022-01-11 09:38:20 ERROR    FBA method file not found
Traceback (most recent call last):
  File "C:\Users\wesle\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-9f2d0322d50f>", line 1, in <module>
    GHG_2016_m1 = flowsa.getFlowBySector("GHG_national_2016_m1")
  File "c:\users\wesle\flowsa\flowsa\__init__.py", line 120, in getFlowBySector
    download_FBAs_if_missing=download_FBAs_if_missing
  File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 134, in main
    flows = load_source_dataframe(k, v, download_FBA_if_missing)
  File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 96, in load_source_dataframe
    download_FBA_if_missing=download_FBA_if_missing)
  File "c:\users\wesle\flowsa\flowsa\__init__.py", line 66, in getFlowByActivity
    flowsa.flowbyactivity.main(year=year, source=datasource)
  File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 195, in main
    config = load_yaml_dict(source, flowbytype='FBA')
  File "c:\users\wesle\flowsa\flowsa\common.py", line 125, in load_yaml_dict
    inherits = config.get('inherits_from')
UnboundLocalError: local variable 'config' referenced before assignment
```

**Comment by catherinebirney** [comment]
> @catherinebirney I made it a lot further this time into the fbs.. then ran into another config related error
> 
> ```
> 2022-01-11 09:38:20 INFO     Retrieving Flow-By-Activity for datasource EPA_GHGI_T_3_42 in year 2016
> 2022-01-11 09:38:20 INFO     EPA_GHGI_T_3_42 2016 not found in C:/Users/wesle/AppData/Local/flowsa/FlowByActivity/, running functions to generate FBA
> 2022-01-11 09:38:20 ERROR    FBA method file not found
> Traceback (most recent call last):
>   File "C:\Users\wesle\AppData\Local\Programs\Python\Python37\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
>     exec(code_obj, self.user_global_ns, self.user_ns)
>   File "<ipython-input-3-9f2d0322d50f>", line 1, in <module>
>     GHG_2016_m1 = flowsa.getFlowBySector("GHG_national_2016_m1")
>   File "c:\users\wesle\flowsa\flowsa\__init__.py", line 120, in getFlowBySector
>     download_FBAs_if_missing=download_FBAs_if_missing
>   File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 134, in main
>     flows = load_source_dataframe(k, v, download_FBA_if_missing)
>   File "c:\users\wesle\flowsa\flowsa\flowbysector.py", line 96, in load_source_dataframe
>     download_FBA_if_missing=download_FBA_if_missing)
>   File "c:\users\wesle\flowsa\flowsa\__init__.py", line 66, in getFlowByActivity
>     flowsa.flowbyactivity.main(year=year, source=datasource)
>   File "c:\users\wesle\flowsa\flowsa\flowbyactivity.py", line 195, in main
>     config = load_yaml_dict(source, flowbytype='FBA')
>   File "c:\users\wesle\flowsa\flowsa\common.py", line 125, in load_yaml_dict
>     inherits = config.get('inherits_from')
> UnboundLocalError: local variable 'config' referenced before assignment
> ```

@WesIngwersen The GHGI FBA is designed to be run only with "EPA_GHGI", so the FBA config for a specific table doesn't exist. I'll address this issue now.

**Comment by WesIngwersen** [comment]
> 

Thanks. @bl-young I assume you had all the FBAs already generated so you didn't run into this

**Comment by catherinebirney** [comment]
@WesIngwersen latest [commit](https://github.com/USEPA/flowsa/pull/165/commits/f0326641306cef95fceffce56dc08b8b57cde6ed) addresses GHGI generation issue

There is an issue with the flow mapping for GHGI which likely has to do with a mismatch in names in flowsa (with the table extensions) and also that the mapping file name is [GHGI](https://github.com/USEPA/Federal-LCA-Commons-Elementary-Flow-List/blob/master/fedelemflowlist/flowmapping/GHGI.csv) instead of EPA_GHGI @bl-young

**Comment by bl-young** [comment]
the mapping file is [specified as GHGI in the FBS config](https://github.com/USEPA/flowsa/blob/ghg_fbs/flowsa/data/flowbysectormethods/GHG_national_2016_m1.yaml#L17).

What's the issue you are seeing? Note that the GHGI mapping in develop (in _fedelemflowlist_) is complete while the one on master is not

**Comment by bl-young** [comment]
@WesIngwersen there have been a lot of changes to the GHGI FBAs. You didn't run into that error for the first FBA listed, which makes me think you could be using an old one. You may want to delete any that you haven't created recently so as to avoid using an old incorrect GHGI FBA

**Comment by catherinebirney** [comment]
> the mapping file is [specified as GHGI in the FBS config](https://github.com/USEPA/flowsa/blob/ghg_fbs/flowsa/data/flowbysectormethods/GHG_national_2016_m1.yaml#L17).
> 
> What's the issue you are seeing? Note that the GHGI mapping in develop (in _fedelemflowlist_) is complete while the one on master is not

@bl-young - thanks switching to the develop branch fixed the mapping

**Comment by WesIngwersen** [comment]
> > the mapping file is [specified as GHGI in the FBS config](https://github.com/USEPA/flowsa/blob/ghg_fbs/flowsa/data/flowbysectormethods/GHG_national_2016_m1.yaml#L17).
> > What's the issue you are seeing? Note that the GHGI mapping in develop (in _fedelemflowlist_) is complete while the one on master is not
> 
> @bl-young - thanks switching to the develop branch fixed the mapping

FYI I was using the latest in develop for all repos, including fedelemflowlist when I ran into the error. I will remove all EPA_GHGI fbas and try again

**Comment by WesIngwersen** [comment]
Just ran into #166 running the fbs when trying to create a CoA Cropland FBA for allocation.

**Comment by bl-young** [comment]
Script for model comparison is on [GHGbyindustrydocs repo](https://github.com/WesIngwersen/GHGbyindustrydocs/blob/main/model_scripts/CompareGHGs.rmd)

**Review by WesIngwersen — APPROVED** [review]
GHG_national_2016_m1 generates successfully with FBAs for QCEW and CoA Cropland already stored.Logs are created successfully.  This approval is not a full review of the GHG_national_2016_m1.

**Pull request merged**